### PR TITLE
Fix eq

### PR DESCRIPTION
**Pull request by dinosaure**
See #48

### CODE REVIEW

**Comment by dinosaure** [comment]
Need to add a test about that.

**Comment by dinosaure** [comment]
I added a test on `digestif` (specially on `digestif.c` but `eq` is common between C and OCaml implementation) and see if 100 computation of `Digestif.eq` take more or less the same time to be executed. Clearly, if people can review this test, I will happy because I'm not really about what I did.

**Comment by dinosaure** [comment]
About `consistent_of_hex` function, I will put in another PR some tests to canonicalize expected behavior of this function.

**Comment by dinosaure** [comment]
Hmmhmm, it seems than the test about `eq` is wrong, may be. The goal is to compute the standard deviation from a series of times needed to compute `eq` on randoms hashes. On my computer, the standard deviation is ~ 13 % (`100 * deviation * (max series - min series)`) and ~ 21 % when I replace `eq` by `String.equal`.

However, from the CI, we have ~ 40 % so the computation is may be wrong. The final goal is to say: from any input, time needed to compute `eq` is, more or less, the same. From outside advises, the percentage show may be noise of the CPU but I don't know how to delete it and if it's the real problem.

So if someone has a idea, a better way or something else to check assumption to avoid timing attacks on `eq` functions, feel free on this PR.

**Comment by cfcs** [comment]
I would expect the new `eq` to be slower than `String.equal` since `String.equal` can return early when they are not equal, and can take advantage of string deduplication?

**Comment by dinosaure** [comment]
Yes your are true but my computation is not about speed but about [standard deviation](https://en.wikipedia.org/wiki/Standard_deviation) from a series of times and, as expected, because `String.equal` can leave earlier (a time needed to be executed is not constant) than `eq`, the variation on this series should be bigger than `eq`.

So results on my computer go in my way: a low standard deviation indicates that the data points tend to be close to the mean. In other words, times registered on the execution of `eq` is more or less the same.

File: test/test_eq.ml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+let hashes = Array.init 1000 (Digestif.SHA1.consistent_of_hex <.> random_hash Digestif.SHA1.digest_size)
+
+let bench n =
+  Bench.Test.create_indexed
+    ~name:"eq"
+    ~args:(list_init n (fun _ -> Random.int 1000))
+  @@ fun i ->
+  let a = Array.get hashes i in
+  let b = Array.get hashes i in
```
Comment (samoht) [inline test/test_eq.ml]: `a` is always equal to `b` here?

Comment (dinosaure) [reply test/test_eq.ml]: Hmmhmm, we does not care about inputs (and it's why it's random inputs) and only time needed to execute `eq` is relevant here.

Comment (samoht) [reply test/test_eq.ml]: but if the input are identical, that means that eq will scan the full bytes anyway.

Comment (dinosaure) [reply test/test_eq.ml]: The implementation should scan full bytes anyway even if `a` and `b` differs. Then I can generate 2 inputs and execute `eq` on them.

**Comment by dinosaure** [comment]
See [@dinosaure/eqaf](https://github.com/dinosaure/eqaf)

**Comment by dinosaure** [comment]
Lol `ocaml.4.07.0` does not compile on Travis CI ...

**Comment by dinosaure** [comment]
PR rebased and cleaned, when Travis will be happy with it, I will merge it.

**Comment by samoht** [comment]
On OSX I've got:

```
$ jbuilder runtest
        test alias test/runtest (exit 2)
(cd _build/default/test && ./test.exe)
Estimated testing time 5m (300 benchmarks x 1s). Change using -quota SECS.
Estimated testing time 5m (300 benchmarks x 1s). Change using -quota SECS.
Got an error: (eqst:90
  "Regression failed. (In Bench, this is usually because the predictors were\
 \nlinearly dependent, commonly as a result of having a predictor that is always\
 \nzero, or having two predictors that are multiples of each other.)").
Uncaught exception:

  (eqst:90
    "Regression failed. (In Bench, this is usually because the predictors were\
   \nlinearly dependent, commonly as a result of having a predictor that is always\
   \nzero, or having two predictors that are multiples of each other.)")

Raised at file "test/test.ml", line 84, characters 10-34
```

**Pull request merged**